docs: Rewrite Keyframe Animations documentation page (#6181)
software-mansion/react-native-reanimated: packages/docs-reanimated/docs/layout-animations/keyframe-animations.mdx

sidebar_position: 2
---

import KeyframeSrc from '!!raw-loader!@site/src/examples/KeyframeAnimation';

# Keyframe animations

> The document explains how you can define complex animation using simple and popular animation definitions schema - Keyframes.

Author: it sounds good I think??

The document explains how you can define complex animation using simple and popular animation definitions schema - Keyframes. They allow you more flexibility than standard Entering and Exiting presets.

> The document explains how you can define complex animation using simple and popular animation definitions schema - Keyframes.

Reviewer: Nowhere in the documentation we introduce the subject. "this document is about blah blah" I think it's a waste of everyone's time.

"It's not an airport, no need to introduce your arrival" 😆 

Lets start with 

Keyframes are... or Keyframes lets you...

> The document explains how you can define complex animation using simple and popular animation definitions schema - Keyframes.

Author: Keyframes are animation definition schemas that let you create complex animations. They allow you more flexibility than standard Entering and Exiting presets.

🤔

> The document explains how you can define complex animation using simple and popular animation definitions schema - Keyframes.

Reviewer: Sounds good

    transform: [{ rotate: '45deg' }],
  },
});

function App() {
  return <Animated.View entering={keyframe} />;
}

> ```

Reviewer 2: I'd add `Type definitions` section under the reference. Users need it to see a complete signature of the function/class etc. You can try looking at, for example, one in [useScrollViewOffset docs page](http://localhost:3000/react-native-reanimated/docs/scroll/useScrollViewOffset)

Author: 🆗 ✅

### Arguments

#### `definitions`

> An object, passed as a first argument that contains definitions of your animation.

Reviewer 2: If we add the `Type definitions` section, then there is no need to specifiy it is a first argument - people will see it clearly 😄 
```suggestion
An object, that contains definitions of your animation.
```

The object keys should be within range `0-100` and correspond to animation progress.

> The object values should consist of style props and/or [easing](/docs/animations/withTiming/#easing) function. If easing property is not provided, it defaults to `Easing.linear`.

Reviewer 2: I'd make the link bigger
```suggestion
The object values should consist of style props and/or [easing function](/docs/animations/withTiming/#easing). If easing property is not provided, it defaults to `Easing.linear`.
```

> - `0` or `from`: Initial state of the object.
> - Middle points (e.g., `45`): Intermediate states of the object.
> - `100` or `to`: Final state of the object.

Reviewer 2: There's no explanation what these 3 points mean tbh
```suggestion
The keys take the following values:
- `0` or `from`: Initial state of the object.
- Middle points (e.g., `45`): Intermediate states of the object.
- `100` or `to`: Final state of the object.
```

> To `0` assign the style, you want for your object at the beginning of the animation and to `100` assign the style you want for your object to have at the end of the animation.

Reviewer 2: This sentence is not only too long, but also highly incomprehensible.
```suggestion
The key `0` (or `from`) should be assigned the style that you want for your object at the beginning of the animation. The key `100` (or `to`) should be assigned the style that you want for your object at the end of the animation.
```

- If you want to animate transform style, make sure that all properties in the transformation array are in the same order in all keyframes.

## Modifiers

```javascript
keyframe
  .duration(1000)
  .delay(500)
  .reduceMotion(ReduceMotion.Never)
  .withCallback((finished) => {
    console.log(`finished without interruptions: ${finished}`);
  });
```

- `.duration(durationMs: number)` is the length of the animation (in milliseconds). Defaults to `500`.
- `.delay(durationMs: number)` is the delay before the animation starts (in milliseconds). Defaults to `0`.
- `.reduceMotion(reduceMotion: ReduceMotion)` determines how the animation responds to the device's reduced motion accessibility setting.
- `.withCallback(callback: (finished: boolean) => void)` is the callback that will fire after the animation ends. Sets `finished` to `true` when animation ends without interruptions, and `false` otherwise.

## Example

<ExampleVideo
  sources={{
    android:
      '/react-native-reanimated/recordings/layout-animations/keyframe_android.mov',
    ios: '/react-native-reanimated/recordings/layout-animations/keyframe_ios.mov',
  }}

> />

Reviewer 2: I'd move the `Modifiers` and `Examples` sections ***before*** the `Remarks`. Remarks are like some additional comments and pitfalls to be aware of, but they assume that the user knows the API, thus I think it is better to have them after Modifiers and Examples.

Author: I left it there because they were talking about the usage of keys, but it's a good idea to move them at the end
